### PR TITLE
perf(python): order(s) of magnitude speedup when initialising `List` dtype `Series` from 2D numpy array

### LINKED ISSUE #12669: Initializing a list column with a numpy 2D-array takes too long
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
import numpy as np

a = np.arange(10_000_000).reshape(-1, 1)  # instant
print(a)
df = pl.DataFrame({"a": a})  # about a minute
print(df)
```

### Log output

```shell
No error
```

### Issue description

Creating a numpy 2D-array takes a few milliseconds on my machine.
But initializing a polars dataframe's column with this array takes more then a minute.

### Expected behavior

Should be instant.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.15
Index type:          UInt32
Platform:            Windows-10-10.0.19041-SP0
Python:              3.10.5 (tags/v3.10.5:f377153, Jun  6 2022, 16:14:13) [MSC v.1929 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
gevent:              <not installed>
matplotlib:          <not installed>
numpy:               1.24.3
openpyxl:            <not installed>
pandas:              2.1.1
pyarrow:             13.0.0
pydantic:            <not installed>
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #12669.

Instead of using a naïve conversion inside a Python loop, we now reshape the input numpy array as 1D, fast-initialise a Series from that, and then defer to a Rust-side reshape operation to optimally recover the given dimensions.

In the specific example given in the issue this results in initialising the frame ~2,500x faster  🚀
(The degree of speedup depends on the dimensionality, but it's typically ~1-3 orders of magnitude).

### Examples

* **Average case**

  ```python
  from codetiming import Timer
  import polars as pl
  import numpy as np

  a = np.arange(1_000_000).reshape(-1, 5)
  with Timer():
      df = pl.DataFrame({"a": a})

  # shape: (200_000, 1)
  # ┌────────────────────────────┐
  # │ a                          │
  # │ ---                        │
  # │ list[i64]                  │
  # ╞════════════════════════════╡
  # │ [0, 1, … 4]                │
  # │ [5, 6, … 9]                │
  # │ …                          │
  # │ [999990, 999991, … 999994] │
  # │ [999995, 999996, … 999999] │
  # └────────────────────────────┘
  ```
  ```
  Before: 0.6246 seconds
   After: 0.0151 seconds
  -------------------------
  Speedup: ~41x 
  -------------------------
  ```

* **Extreme case**
  ```python
  a = np.arange(100_000_000).reshape(-1, 1)
  with Timer():
      df = pl.DataFrame({"a": a})

  # shape: (100_000_000, 1)
  # ┌────────────┐
  # │ a          │
  # │ ---        │
  # │ list[i64]  │
  # ╞════════════╡
  # │ [0]        │
  # │ [1]        │
  # │ …          │
  # │ [99999998] │
  # │ [99999999] │
  # └────────────┘
  ```
  ```
  Before: 349.3047 seconds
   After:   0.0738 seconds
  -------------------------
  Speedup: ~4,700x 
  -------------------------
  ```

---

### Also

* Splits up the overly-long `test_init_ndarray` func into a few more manageable pieces.
* Fixes an edge-case (ref: #10725) when initialising square numpy arrays; can resolve an otherwise ambiguous tiebreak by checking the flags to confirm c-contiguous vs fortran row/col major order.

### CODE REVIEW

**Comment by alexander-beedie** [comment]
(Will fix the test dtype in the morning! ;)

**Comment by MarcoGorelli** [comment]
Only ~2,500x faster?

**Comment by alexander-beedie** [comment]
> Only ~2,500x faster?

I'll try harder 🙏

File: py-polars/polars/utils/_construction.py [hunk 1]
```diff
@@ -225,15 +225,22 @@ def numpy_to_pyseries(
...
-    elif len(values.shape) == 2:
-        pyseries_container = []
-        for row in range(values.shape[0]):
-            pyseries_container.append(
-                numpy_to_pyseries(
-                    "", values[row, :], strict=strict, nan_to_null=nan_to_null
-                )
-            )
-        return PySeries.new_series_list(name, pyseries_container, _strict=False)
+    elif len(shape := values.shape) == 2:
```
Comment (ritchie46) [inline py-polars/polars/utils/_construction.py:228]: The famous walrus operator! ^^

File: py-polars/polars/utils/_construction.py [hunk 2]
```diff
@@ -225,15 +225,22 @@ def numpy_to_pyseries(
         return constructor(
             name, values, nan_to_null if dtype in (np.float32, np.float64) else strict
         )
-    elif len(values.shape) == 2:
-        pyseries_container = []
-        for row in range(values.shape[0]):
```
Comment (ritchie46) [inline py-polars/polars/utils/_construction.py:230]: Ouch, that hurts indeed.

**Review by ritchie46 — APPROVED** [review]

**Comment by ritchie46** [comment]
Really nice!

**Comment by zaa730Sight** [comment]
Any reason we don't default this kind of 2d np -> pl behavior instead?

```python
import polars as pl
import numpy as np
import time

t1 = time.time_ns()

a = np.arange(1000000000).reshape(-1, 100)

t2 = time.time_ns()

df_fast = pl.DataFrame({f"column_{i}": a[:, i] for i in range(100)})

t3 = time.time_ns()

df_slow = pl.DataFrame(a)

t4 = time.time_ns()

arr_alloc = round((t2 - t1) / 1e9, 2)
df_fast_time = round((t3 - t2) / 1e9, 2)
df_slow_time = round((t4 - t3) / 1e9, 2)

print(f"Time to create array: {arr_alloc} s, fast df: {df_fast_time} s, slow df: {df_slow_time} s")

print(df_fast.equals(df_slow)) # True

```

I got "Time to create array: 1.41 s, fast df: 0.37 s, slow df: 10.6 s" on my humble machine

```python
pl.show_versions()
--------Version info---------
Polars:              0.19.16
Index type:          UInt32
Platform:            Linux-5.15.0-1027-aws-x86_64-with-glibc2.31
Python:              3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.1
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.9.1
gevent:              <not installed>
matplotlib:          3.7.2
numpy:               1.26.1
openpyxl:            3.1.2
pandas:              2.1.3
pyarrow:             14.0.1
pydantic:            2.4.2
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          1.4.46
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

```bash
(base) cpcv@r7i24xlarge-dy-e7-1:~$ lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-47
Off-line CPU(s) list:            48-95
Thread(s) per core:              1
Core(s) per socket:              48
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           143
Model name:                      Intel(R) Xeon(R) Platinum 8488C
Stepping:                        8
CPU MHz:                         3074.477
BogoMIPS:                        4800.00
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       2.3 MiB
L1i cache:                       1.5 MiB
L2 cache:                        96 MiB
L3 cache:                        105 MiB
NUMA node0 CPU(s):               0-47
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequenc
                                 e
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush
                                 mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon r
                                 ep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq
                                  monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline
                                 _timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_si
                                 ngle ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2
                                 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512c
                                 d sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 w
                                 bnoinvd ida arat avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmu
                                 lqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid cldemote movdiri movdi
                                 r64b md_clear serialize amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_ca
                                 pabilities

```

**Pull request merged**